nemo/collections/tts/modules/transformer_2412.py

```diff
@@ -0,0 +1,700 @@
+                query=x_normed, query_mask=x_mask, memory=memory, memory_mask=cond_mask, attn_prior=attn_prior
+            )
+            if self.use_cache:
+                if self.cache['cross_attn_output'] is not None:
+                    x_res = torch.cat([self.cache['cross_attn_output'], x_res], dim=1)
+                self.cache['cross_attn_output'] = x_res
+            x = x + x_res
+
+        # mlp final projection
+        x = x + self.pos_ff(self.norm_pos_ff(x))
+        x = x * x_mask.unsqueeze(-1)
+
+        return {
+            'output': x,
+            'attn_probabilities': {'self_attn_probabilities': s_attn_prob, 'cross_attn_probabilities': x_attn_prob},
+        }
+
+
+class Transformer(torch.nn.Module):
+    def __init__(
+        self,
+        n_layers: int,
+        d_model: int,
+        d_ffn: int,
+        sa_n_heads: int,
+        kernel_size: int,
+        p_dropout: float = 0.0,
+        p_dropout_out: float = 0.0,
+        has_xattn: bool = False,
+        xa_d_memory: Optional[int] = None,
+        xa_n_heads: Optional[int] = None,
+        is_causal: bool = True,
+        apply_norm_to_cond: bool = True,
+        apply_norm_out: bool = False,
+        max_length_causal_mask: int = 4096,
+        use_learnable_pos_emb: bool = False,
+        conv_non_linearity: Callable = torch.nn.GELU(approximate="tanh"),
+    ):
+        """
+        Initializes a stack of transformer layers. Can be used for both encoder and decoder.
+        Set is_causal is True for autoregressive models. Equivalent to TransformerBlock from Megatron-LM
+        Args:
+            n_layers <int>: Number of transformer layers
+            d_model <int>: Model dimension
+            d_ffn <int>: Feed forward dimension (usually 4*d_model)
+            sa_n_heads <int>: Number of attention heads used in self-attention
+            kernel_size <int>: Convolution kernel size for FFN
+            p_dropout <float>: Dropout probability
+            p_dropout_out <float>: Dropout probability for output
+            has_xattn <bool>: Whether to use cross attention
+            xa_d_memory <int>: Hidden dimension for cross attention; required if has_xattn is True
+            xa_n_heads <int>: Number of attention heads used in cross attention; required if has_xattn is True
+            is_causal <bool>: Whether to make attention and the convolution feedforward networks causal.
+            apply_norm_to_cond <bool>: Whether to apply normalization to conditioning tensor; conditioning tensor being
+                the input to the memory part of cross-attention.
+            apply_norm_out <bool>: Whether to apply normalization to output
+            max_length_causal_mask <int>: Maximum length of causal mask
+            use_learnable_pos_emb <bool>: Whether to add a learnable positionable embedding inside the class
+            conv_non_linearity <Callable>: Convolution non-linearity
+        """
+        super().__init__()
```

Comment:
need a safeguard, otherwise, fail,
```suggestion
        if has_xattn and (xa_d_memory is None or xa_n_heads is None):
            raise ValueError("It requires that `xa_d_memory` and `xa_n_heads` are specified when `has_xattn` is True!")
        super().__init__()
```